app/src/main/java/org/simple/clinic/drugs/selection/custom/CustomDrugEntrySheet.kt

```diff
@@ -0,0 +1,186 @@
+import org.simple.clinic.navigation.v2.fragments.BaseBottomSheet
+import org.simple.clinic.util.unsafeLazy
+import org.simple.clinic.widgets.ScreenDestroyed
+import org.simple.clinic.widgets.UiEvent
+import org.simple.clinic.widgets.textChanges
+import java.util.Locale
+import java.util.UUID
+import javax.inject.Inject
+
+class CustomDrugEntrySheet : BaseBottomSheet<
+    CustomDrugEntrySheet.Key,
+    SheetCustomDrugEntryBinding,
+    CustomDrugEntryModel,
+    CustomDrugEntryEvent,
+    CustomDrugEntryEffect,
+    Unit>(), CustomDrugEntryUi, CustomDrugEntrySheetUiActions {
+
+  @Inject
+  lateinit var locale: Locale
+
+  @Inject
+  lateinit var features: Features
+
+  @Inject
+  lateinit var router: Router
+
+  @Inject
+  lateinit var effectHandlerFactory: CustomDrugEntryEffectHandler.Factory
+
+  private val openAs by unsafeLazy { screenKey.openAs }
+
+  private val titleTextView
+    get() = binding.titleTextView
+
+  private val removeButton
+    get() = binding.removeButton
+
+  private val drugDosageEditText
+    get() = binding.drugDosageEditText
+
+  private val drugFrequencyEditText
+    get() = binding.drugFrequencyEditText
+
+  private val saveButton
+    get() = binding.saveButton
+
+  override fun defaultModel() = CustomDrugEntryModel.default(openAs)
+
+  override fun bindView(
+      inflater: LayoutInflater,
+      container: ViewGroup?
+  ) = SheetCustomDrugEntryBinding.inflate(inflater, container, false)
+
+  override fun uiRenderer() = CustomDrugEntryUiRenderer(this, getString(R.string.custom_drug_entry_sheet_dosage_placeholder))
+
+  override fun createUpdate() = CustomDrugEntryUpdate()
+
+  override fun createInit() = CustomDrugEntryInit()
+
+  override fun createEffectHandler(viewEffectsConsumer: Consumer<Unit>) = effectHandlerFactory.create(this).build()
+
+  override fun events() = Observable
+      .mergeArray(
+          drugDosageChanges(),
+          drugDosageFocusChanges(),
+          saveClicks(),
+          removeClicks()
+      ).compose(ReportAnalyticsEvents())
+      .cast<CustomDrugEntryEvent>()
+
+  override fun onAttach(context: Context) {
+    super.onAttach(context)
+
+    context.injector<Injector>().inject(this)
+  }
+
+  private fun drugDosageChanges() = drugDosageEditText.textChanges(::DosageEdited)
+
+  private fun drugDosageFocusChanges() = drugDosageEditText.focusChanges().map(::DosageFocusChanged)
+
+  private fun saveClicks(): Observable<UiEvent> {
+    val dosageImeClicks = drugDosageEditText
+        .editorActions { it == EditorInfo.IME_ACTION_DONE }
+        .map { Unit }
+
+    return saveButton
+        .clicks()
+        .mergeWith(dosageImeClicks)
+        .map { AddMedicineButtonClicked(screenKey.patientUuid) }
+  }
+
+  private fun removeClicks(): Observable<UiEvent> =
+      removeButton
+          .clicks()
+          .map { RemoveDrugButtonClicked }
+
+  override fun showEditFrequencyDialog(frequency: DrugFrequency) {
+    // to do
+  }
+
+  @Suppress("IMPLICIT_CAST_TO_ANY")
```

Comment:
Why are we suppressing this? 🤔

Reply:
It was giving me a waning earlier. It's not anymore 😅